examples/mnit_ballz/src/game_logic.nit

```diff
@@ -14,167 +14,172 @@
+			return bounce(center, intersecting_line.point_left, intersecting_line.point_right, offset)
 		end
+		return null
 	end
 
-	fun activate_walls(event: MotionEvent)
-	do
-		if event.just_went_down then
-			walls_activated = not walls_activated
-		end
-	end
+	# Event interception
+	fun intercepts(event: InputEvent): Bool is abstract
 end
 
-class Screen
-	var ball_img: Image
-	var game: Game
-
-	init(app: App, display: Display)
-	do
-		game = new Game(display)
-		ball_img = app.load_asset("images/ball.png").as(Image)
-		var scale = game.img_dim.to_f / game.img_ori_dim.to_f
-		ball_img.scale = scale
-		ball_img.scale = 3.0
-	end
+# The core of the game
+class Game
 
-	fun do_frame(display: Display)
-	do
-		display.clear(0.0, 0.0, 0.0)
-		display.blit_rotated(ball_img, game.ball.x, game.ball.y, 0.0)
-	end
+	# The Ball!
+	var ball: Ball is noinit
 
-	fun input(ie: InputEvent): Bool
-	do
-		if ie isa ASensorProximity then
-			if ie.distance == 0.0 then ball_img.scale = 6.0 else ball_img.scale = 3.0
-		else
-			game.ball.intercepts(ie)
-		end
-		return true
-	end
-end
+	# List of walls in the level
+	var walls: Array[Wall] is noinit
 
-class Game
-	var ball: Ball
+	# Width of the display
 	var width: Float
+
+	# Heightof the display
 	var height: Float
 
-	var img_ori_dim: Int = 256
-	fun img_dim: Int do return 210
+	# Quadtree used for collision detection
+	var quadtree: SQuadTree[OrientedLine] is noinit
 
-	init(display: Display)
+	init
 	do
-		width = display.width.to_f
-		height = display.height.to_f
-		ball = new Ball(self, width/2.0, height/2.0, false)
+		ball = new Ball(new Point[Float](width/2.0, height/2.0))
+		# Walls initialisation
+		var walla = new Wall(new Point[Float](width/4.0, height/4.0), pi/3.0, 1.0)
+		var wallb = new Wall(new Point[Float](width*0.75, height/4.0), 0.0, 1.0)
+		var wallc = new Wall(new Point[Float](width/4.0, height*0.75), 0.0, 1.0)
+		var walld = new Wall(new Point[Float](width*0.75, height*0.75), pi/3.0, 1.0)
+		walls = new Array[Wall].with_items(walla, wallb, wallc, walld)
+
+		# adding screen bordures
+		var i = new Point[Float](0.0,0.0)
+		var a = new Point[Float](0.0, height/2.0)
+		var b = new Point[Float](width/2.0, 0.0)
+		var c = new Point[Float](width, height/2.0)
+		var d = new Point[Float](width/2.0, height)
+
+		var l1 = new OrientedLine(i, i, pi/2.0, height, a)
+		var l2 = new OrientedLine(i, i, 0.0, width, b)
+		var l3 = new OrientedLine(i, i, pi/2.0, height, c)
+		var l4 = new OrientedLine(i, i, 0.0, width, d)
+
+		quadtree = new SQuadTree[OrientedLine](5, width, height)
+		for w in walls do for l in w.lines do
+			quadtree.add(l)
+		end
+		quadtree.add(l1)
+		quadtree.add(l2)
+		quadtree.add(l3)
+		quadtree.add(l4)
 	end
 
-	fun do_turn
-	do
-	ball.do_turn
+	# Only calls `do_turn` of the ball for the moment
+	fun do_turn do ball.do_turn(self)
+
+	# Input gestion
+	fun input(ie: InputEvent): Bool do return false
+end
+
+redef class App
+
+	# Assets used in all the app
+	var assets = new Assets
```

Comment:
logic issue here: `new Assets` will require things from `app` but because you are in the default value of an attribute of App, the object is not initialized yet thus `app` is not assigned. Therefore that will do an infinite recursion that will segfault the program.